### PR TITLE
chore(gatsby): expose full structured logs over the develop status websocket server

### PR DESCRIPTION
**Pull request by mxstbr**
In order to make the prompt-to-restart work (#22759), we added a websocket server to the develop parent process that emits certain events. At the time, we decided to make them specific to the use case and have events such as "develop:needs-restart" when certain IPC messages arrived from the child process.

This PR changes it to emit a "structed-log" event that forwards _all_ structured logs. We will need this anyway to show the logs in Admin, and it's cleaner (less custom stuff and more reusability === better!). It also introduces a new structured log that is exclusive to the websocket server to note that the develop process needs to be restarted which looks like this:

```JS
{
  type: "LOG_ACTION",
  payload: {
    type: "DEVELOP",
    payload: "RESTART_REQUIRED"
  }
}
```

I don't think the naming is perfect (`type: "DEVELOP"` what's that?!) but it should be good enough. If anybody has better ideas for the names I'm all ears!

### CODE REVIEW

**Comment by mxstbr** [comment]
Canary published for testing on Cloud:

```
Sucessfully published:
 - gatsby-admin@0.1.69-structured-logs-websocket.564+ef0f30353
 - gatsby@2.23.5-structured-logs-websocket.564+ef0f30353
```

**Comment by mxstbr** [comment]
Seems to work fine on Cloud in Preview! Thanks to @wardpeet for testing that with me :pray:

File: packages/gatsby/src/commands/develop.ts
Comment (wardpeet) [inline packages/gatsby/src/commands/develop.ts:304]: Could you explain this? I'm not super familiar with socket.io and unsure what ackFn does here. Does it send a new message? Maybe worth a comment in code?

Comment (mxstbr) [reply packages/gatsby/src/commands/develop.ts:304]: It sends back an acknowledgement, kinda like a response to a HTTP request! On the client, this allows us to listen for it like so:

```JS
socket.emit(`develop:restart`, (answer) => {
  console.log(answer) // => would be "develop:is-starting" here
})
```

I'll add a comment 👍

Comment (mxstbr) [reply packages/gatsby/src/commands/develop.ts]: Added a comment _and_ renamed the variable to `respond` for clarity 👍

File: packages/gatsby/src/utils/restarting-screen.ts
Comment (pieh) [inline packages/gatsby/src/utils/restarting-screen.ts:104]: What should happen if it's not success? (it can be `FAILED` for multiple reasons - this could some failure when doing webpack compilation, or during query extraction).

Also - we emit this potentially multiple times after starting `gatsby develop`, so this  could mean that each webpack compilation (hot reloads) actually force hard reload I think

Comment (mxstbr) [reply packages/gatsby/src/utils/restarting-screen.ts:104]: We should add an error state to this screen, good call! Let's do this after this PR as that's the current behavior as well and I don't want this branch to be forever long lived :grimacing:

Comment (mxstbr) [reply packages/gatsby/src/utils/restarting-screen.ts:104]: #26064

File: packages/gatsby/cache-dir/app.js
Comment (pvdz) [inline packages/gatsby/cache-dir/app.js:49]: Nit: `window.confirm` is blocking. You could put this in the same conditional block and drop the extra indentation block... :shrug:

File: packages/gatsby/cache-dir/app.js
Comment (pvdz) [inline packages/gatsby/cache-dir/app.js:57]: How does the page recover / notify the user if an error happens after `.emit`, causing the callback not to be triggered? 

Maybe a `setTimeout` can be used here as a heartbeat kind of thing? It wouldn't need to be cleared because the `.reload` would do that implicitly. It would allow us to give a cue to the user to recover manually, somehow, rather than beachballing them :beach_umbrella:

**Comment by pvdz** [comment]
Is it very hard to have some kind of sanity test for this stuff? `.reload()` stuff is notoriously difficult to track and test.

File: packages/gatsby/cache-dir/app.js
Comment (mxstbr) [reply packages/gatsby/cache-dir/app.js]: Great call! Let's do that as part of #26064!

File: e2e-tests/development-runtime/cypress.json
Comment (mxstbr) [inline e2e-tests/development-runtime/cypress.json:4]: Note: I had to enable this, otherwise the `window.location.reload()` calls in the restarting flow break Crypess. :confused:

**Comment by mxstbr** [comment]
I wrote a beautiful e2e test that covers the happy path restarting flow that runs perfectly locally—and breaks in CI without any indication as to why :grimacing:

**Comment by mxstbr** [comment]
I could not figure out how to un-flake the e2e test, so I commented it out for now to unblock this PR from shipping. (the e2e test didn't exist before this PR anyway)

File: e2e-tests/development-runtime/cypress.json
Comment (laurieontech) [reply e2e-tests/development-runtime/cypress.json:4]: Crypess lol, amazing typo

File: packages/gatsby/cache-dir/app.js
Comment (laurieontech) [inline packages/gatsby/cache-dir/app.js:45]: NB: Is it worth making these constants up top in case we need to change the business logic?

**Review by laurieontech — APPROVED** [review]
Looks reasonable to me though I'm not super familiar with websockets, so might want a second pair of eyes.

**Pull request merged**
